README.md
# CodePush Server [source](https://github.com/lisong/code-push-server) 

[![NPM](https://nodei.co/npm/code-push-server.svg?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/code-push-server/)

[![NPM Version](https://img.shields.io/npm/v/code-push-server.svg)](https://npmjs.org/package/code-push-server)
[![Node.js Version](https://img.shields.io/node/v/code-push-server.svg)](https://nodejs.org/en/download/)
[![Linux Status](https://img.shields.io/travis/lisong/code-push-server/master.svg?label=linux)](https://travis-ci.org/lisong/code-push-server)
[![Windows Status](https://img.shields.io/appveyor/ci/lisong/code-push-server/master.svg?label=windows)](https://ci.appveyor.com/project/lisong/code-push-server)
[![Coverage Status](https://img.shields.io/coveralls/lisong/code-push-server/master.svg)](https://coveralls.io/github/lisong/code-push-server)
[![Dependency Status](https://img.shields.io/david/lisong/code-push-server.svg)](https://david-dm.org/lisong/code-push-server)
[![Known Vulnerabilities](https://snyk.io/test/npm/code-push-server/badge.svg)](https://snyk.io/test/npm/code-push-server)
[![Licenses](https://img.shields.io/npm/l/code-push-server.svg)](https://spdx.org/licenses/MIT)

CodePush Server is a CodePush progam server! microsoft CodePush cloud is slow in China, we can use this to build our's. I use [qiniu](http://www.qiniu.com/) to store the files, because it's simple and quick!  Or you can use [local/s3/oss/tencentcloud] storage, just modify config.js file, it's simple configure.


## Support Storage mode 

- local *storage bundle file in local machine*
- qiniu *storage bundle file in [qiniu](http://www.qiniu.com/)*
- s3 *storage bundle file in [aws](https://aws.amazon.com/)*
- oss *storage bundle file in [aliyun](https://www.aliyun.com/product/oss)*
- tencentcloud *storage bundle file in [tencentcloud](https://cloud.tencent.com/product/cos)*

## qq Exchange Group

-QQ Group: 628921445
-QQ Group: 535491067

## Correctly use code-push hot update

-Apple App allows hot updates [Apple's developer agreement] (https://developer.apple.com/programs/ios/information/iOS_Program_Information_4_3_15.pdf), in order to not affect the user experience, it is required to use silent updates. Google Play cannot use silent updates. You must pop the box to inform users that the App has an update. The android market in China must adopt a silent update (if the pop-up box prompts, the app will be rejected by "Please upload the latest version of the binary application package").
-react-native Bundle packages are different for different platforms. When using code-push-server, different applications must be created to distinguish them (eg. CodePushDemo-ios and CodePushDemo-android)
-react-native-code-push only updates the resource files, not java and Objective C, so when npm upgrades the dependent package version, if the localized implementation used by the dependent package, you must change the application version number (ios modify Info CFBundleShortVersionString in .plist, android modify the versionName in build.gradle), and then recompile the app and publish it to the app store.
-It is recommended to use the code-push release-react command to release the application, which combines the package and release commands (eg. Code-push release-react CodePushDemo-ios ios -d Production)
-Every time you submit a new version to the App Store, you should also release an initial version to code-push-server based on the submitted version. (Because every time a version is released to code-push-server later, code-puse-server will be compared with the initial version to generate a patch version)


### shell login

```shell
$ code-push login http://api.code-push.com #登录
```

### [web](http://www.code-push.com) 

访问：http://www.code-push.com

### client eg.

[ReactNative CodePushDemo](https://github.com/lisong/code-push-demo-app)

[Cordova CodePushDemo](https://github.com/lisong/code-push-cordova-demo-app)

## HOW TO INSTALL code-push-server

- [docker](https://github.com/lisong/code-push-server/blob/master/docker/README.md) (recommend)
- [manual operation](https://github.com/lisong/code-push-server/blob/master/docs/README.md)

## DEFAULT ACCOUNT AND PASSWORD

- account: `admin`
- password: `123456`

## HOW TO USE

- [normal](https://github.com/lisong/code-push-server/blob/master/docs/react-native-code-push.md)
- [react-native-code-push](https://github.com/Microsoft/react-native-code-push)
- [code-push](https://github.com/Microsoft/code-push)


## ISSUES

[code-push-server normal solution](https://github.com/lisong/code-push-server/issues/135)

[An unknown error occurred](https://github.com/lisong/code-push-server/issues?utf8=%E2%9C%93&q=unknown)

[modify password](https://github.com/lisong/code-push-server/issues/43)


# UPDATE TIME LINE

- targetBinaryVersion support
  - `*` 
  - `1.2.3`
  - `1.2`/`1.2.*`
  - `1.2.3 - 1.2.7`
  - `>=1.2.3 <1.2.7`
  - `~1.2.3`
  - `^1.2.3`


## Advance Feature

> use google diff-match-patch calculate text file diff patch

- support iOS and Android
- use `"react-native-code-push": "git+https://git@github.com/lisong/react-native-code-push.git"` instead `"react-native-code-push": "x.x.x"` in `package.json`
- change `apps`.`is_use_diff_text` to `1` in mysql codepush database

## License
MIT License [read](https://github.com/lisong/code-push-server/blob/master/LICENSE)




CHANGELOG.md
# Changelog for code-push-server

## 0.5.x

## New Features
-Optimized for incremental text updates, using google `diff-match-patch` algorithm to calculate differences
   -react-native-code-push Android client adaptation, you need to merge https://github.com/Microsoft/react-native-code-push/pull/1393 to use the incremental text update function normally.
  -react-native-code-push iOS client adaptation (need to merge https://github.com/Microsoft/react-native-code-push/pull/1399)
  -react-native-code-push Windows client adaptation (in progress)

## fixbug

-Fix statistics activations
-Fix grayscale release bug
-Added calculation after rollback and last incremental update version

## How to upgrade to this version

### Upgrade database

`$ npm run upgrade`

or

`$ code-push-server-db upgrade`


## 0.4.x

### New Features

-targetBinaryVersion supports regular matching, `deployments_versions` new fields` min_version`, `max_version`
  -`*` Matches all versions
  -`1.2.3` matches specific version` 1.2.3`
  -`1.2` /` 1.2. * `Matches all 1.2 patch versions
  -`> = 1.2.3 <1.3.7`
  -`~ 1.2.3` matches`> = 1.2.3 <1.3.0`
  -`^ 1.2.3` matches`> = 1.2.3 <2.0.0`
-Add docker orchestration service deployment, update documentation
-Support Tencent cloud cos storageType

## How to upgrade to this version

-Upgrade database
`$ ./bin/db upgrade`
or
`$ mysql codepush <. / sql / codepush-v0.4.0-patch.sql`

-Handling stock data
`` `shell
   $ git clone https://github.com/lisong/tools
   $ cd tools
   $ npm i
   $ vim ./bin/fixMinMaxVersion // Modify data configuration
   $ node ./bin/fixMinMaxVersion // success message appears
`` `

## 0.3.x

-Support grayscale publishing
-Adapted the `code-push app add` command, applications no longer distinguish platforms by name, but platforms by type
  -Database table apps add new fields `os`,` platform`
-Improve `code-push release / release-react / release-cordova` command
  -Added `is_disabled` and` rollout` fields in database table packages
-Adapted to `code-push patch` command
-Added `log_report_download`,` log_report_deploy` log tables
-Upgrade npm dependencies

config/config.test.js
var os = require('os');

var config = {};
config.test = {
  db: {
    username: "root",
    password: null,
    database: "codepush_test",
    host: "127.0.0.1",
    port: 3306,
    dialect: "mysql",
    logging: false,
    operatorsAliases: false,
  },
  local: {
    storageDir: os.tmpdir(),
    downloadUrl: "http://127.0.0.1:9027/download",
    public: '/download'
  },
  jwt: {
    tokenSecret: 'INSERT_RANDOM_TOKEN_KEY'
  },
  common: {
    tryLoginTimes: 10,
    diffNums: 3,
    dataDir: os.tmpdir(),
    storageType: "local",
    updateCheckCache: true,
    rolloutClientUniqueIdCache: false,
  },
  smtpConfig: false,
  redis: {
    default: {
      host: "127.0.0.1",
      port: 6379,
      retry_strategy: function (options) {
        if (options.error.code === 'ECONNREFUSED') {
          return new Error('The server refused the connection');
        }
        if (options.total_retry_time > 1000 * 60 * 60) {
            return new Error('Retry time exhausted');
        }
        if (options.times_connected > 10) {
            return undefined;
        }
        return Math.max(options.attempt * 100, 3000);
      }
    }
  }
}
config.test.log4js = {
  appenders: {console: { type: 'console'}},
  categories : {
    "default": { appenders: ['console'], level:'error'},
    "startup": { appenders: ['console'], level:'info'},
    "http": { appenders: ['console'], level:'info'}
  }
}
module.exports = config;


core/const.js
function define(name, value) {
    Object.defineProperty(exports, name, {
        value:      value,
        enumerable: true
    });
}

// Define supported platforms
define("IOS", 1);
define("IOS_NAME", 'iOS');
define("ANDROID", 2);
define("ANDROID_NAME", 'Android');
define("WINDOWS", 3);
define("WINDOWS_NAME", 'Windows');

// define supported application types
define("REACT_NATIVE", 1);
define("REACT_NATIVE_NAME", 'React-Native');
define("CORDOVA", 2);
define("CORDOVA_NAME", 'Cordova');

define("PRODUCTION", 'Production');
define("STAGING", 'Staging');


define("IS_MANDATORY_YES", 1);
define("IS_MANDATORY_NO", 0);


define("IS_DISABLED_YES", 1);
define("IS_DISABLED_NO", 0);


define("RELEAS_EMETHOD_PROMOTE", 'Promote');
define("RELEAS_EMETHOD_UPLOAD", 'Upload');

define("DEPLOYMENT_SUCCEEDED", 1);
define("DEPLOYMENT_FAILED", 2);

define("DIFF_MANIFEST_FILE_NAME", 'hotcodepush.json');

// whether the text file uses google diff-match-patch to calculate the difference
define("IS_USE_DIFF_TEXT_NO", 0);
define("IS_USE_DIFF_TEXT_YES", 1);


define("CURRENT_DB_VERSION", '0.5.0');

routes/users.js
var express = require('express');
var router = express.Router();
var Promise = require('bluebird');
var _ = require('lodash');
var models = require('../models');
var middleware = require('../core/middleware');
var AccountManager = require('../core/services/account-manager');
var AppError = require('../core/app-error')

router.get('/', middleware.checkToken, (req, res) => {
  res.send({ title: 'CodePushServer' });
});

router.post('/', (req, res, next) => {
  var email = _.trim(_.get(req, 'body.email'));
  var token = _.trim(_.get(req, 'body.token'));
  var password = _.trim(_.get(req, 'body.password'));
  var accountManager = new AccountManager();
  return accountManager.checkRegisterCode(email, token)
  .then((u) => {
    if (_.isString(password) && password.length < 6) {
      throw new AppError.AppError('Please enter a password of 6 to 20 characters');
    }
    return accountManager.register(email, password);
  })
  .then(() => {
    res.send({status: "OK"});
  })
  .catch((e) => {
    if (e instanceof AppError.AppError) {
      res.send({status: "ERROR", message: e.message});
    } else {
      next(e);
    }
  });
});

router.get('/exists', (req, res, next) => {
  var email = _.trim(_.get(req, 'query.email'));
  models.Users.findOne({where: {email: email}})
  .then((u) => {
    if (!email) {
      throw new AppError.AppError(`请您输入邮箱地址`);
    }
    res.send({status: "OK", exists: u ? true : false});
  })
  .catch((e) => {
    if (e instanceof AppError.AppError) {
      res.send({status: "ERROR", message: e.message});
    } else {
      next(e);
    }
  });
});

router.post('/registerCode', (req, res, next) => {
  var email = _.get(req, 'body.email');
  var accountManager = new AccountManager();
  return accountManager.sendRegisterCode(email)
  .then(() => {
    res.send({status: "OK"});
  })
  .catch((e) => {
    if (e instanceof AppError.AppError) {
      res.send({status: "ERROR", message: e.message});
    } else {
      next(e);
    }
  });
});

router.get('/registerCode/exists', (req, res, next) => {
  var email = _.trim(_.get(req, 'query.email'));
  var token = _.trim(_.get(req, 'query.token'));
  var accountManager = new AccountManager();
  return accountManager.checkRegisterCode(email, token)
  .then(() => {
    res.send({status: "OK"});
  })
  .catch((e) => {
    if (e instanceof AppError.AppError) {
      res.send({status: "ERROR", message: e.message});
    } else {
      next(e);
    }
  });
});

//修改密码
router.patch('/password', middleware.checkToken, (req, res, next) => {
  var oldPassword = _.trim(_.get(req, 'body.oldPassword'));
  var newPassword = _.trim(_.get(req, 'body.newPassword'));
  var uid = req.users.id;
  var accountManager = new AccountManager();
  return accountManager.changePassword(uid, oldPassword, newPassword)
  .then(() => {
    res.send({status: "OK"});
  })
  .catch((e) => {
    if (e instanceof AppError.AppError) {
      res.send({status: "ERROR", message: e.message});
    } else {
      next(e);
    }
  });
});

module.exports = router;


config/config.js
var os = require('os');

var config = {};
config.development = {
  // Config for database, only support mysql.
  db: {
    username: process.env.RDS_USERNAME || "root",
    password: process.env.RDS_PASSWORD || "welcome",
    database: process.env.DATA_BASE || "codepush",
    host: process.env.RDS_HOST || "localhost",
    port: process.env.RDS_PORT || 3306,
    dialect: "mysql",
    logging: false,
    operatorsAliases: false,
  },
  // Config for qiniu (http://www.qiniu.com/) cloud storage when storageType value is "qiniu".
  qiniu: {
    accessKey: "",
    secretKey: "",
    bucketName: "",
    downloadUrl: "" // Binary files download host address.
  },
  // Config for upyun (https://www.upyun.com/) storage when storageType value is "upyun"
  upyun: {
    storageDir: process.env.UPYUN_STORAGE_DIR,
    serviceName: process.env.UPYUN_SERVICE_NAME,
    operatorName: process.env.UPYUN_OPERATOR_NAME,
    operatorPass: process.env.UPYUN_OPERATOR_PASS,
    downloadUrl: process.env.DOWNLOAD_URL,
  },
  // Config for Amazon s3 (https://aws.amazon.com/cn/s3/) storage when storageType value is "s3".
  s3: {
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
    sessionToken: process.env.AWS_SESSION_TOKEN, //(optional)
    bucketName: process.env.BUCKET_NAME,
    region: process.env.REGION,
    downloadUrl: process.env.DOWNLOAD_URL, // binary files download host address.
  },
  // Config for Aliyun OSS (https://www.aliyun.com/product/oss) when storageType value is "oss".
  oss: {
    accessKeyId: "",
    secretAccessKey: "",
    endpoint: "",
    bucketName: "",
    prefix: "", // Key prefix in object key
    downloadUrl: "", // binary files download host address.
  },
  // Config for tencentyun COS (https://cloud.tencent.com/product/cos) when storageType value is "oss".
  tencentcloud: {
    accessKeyId: "",
    secretAccessKey: "",
    bucketName: "",
    region: "",
    downloadUrl: "", // binary files download host address.
  },
  // Config for local storage when storageType value is "local".
  local: {
    // Binary files storage dir, Do not use tmpdir and it's public download dir.
    storageDir: process.env.STORAGE_DIR || "/Users/jakki/projects/trisysit-poc/code-push-server/workspaces/storage",
    // Binary files download host address which Code Push Server listen to. the files storage in storageDir.
    downloadUrl: process.env.LOCAL_DOWNLOAD_URL || "http://192.168.1.6:9027/download",
    // public static download spacename.
    public: '/download'
  },
  jwt: {
    // Recommended: 63 random alpha-numeric characters
    // Generate using: https://www.grc.com/passwords.htm
    tokenSecret: process.env.TOKEN_SECRET ||'EB0B2CAC8C51CF20B103B0B53CF7E2258088BAF6C417A82C2537F66CE6DC5D27'
  },
  common: {
    /*
     * tryLoginTimes is control login error times to avoid force attack.
     * if value is 0, no limit for login auth, it may not safe for account. when it's a number, it means you can
     * try that times today. but it need config redis server.
     */
    tryLoginTimes: 0,
    // CodePush Web(https://github.com/lisong/code-push-web) login address.
    //codePushWebUrl: "http://127.0.0.1:3001/login",
    // create patch updates's number. default value is 3
    diffNums: 3,
    // data dir for caclulate diff files. it's optimization.
    dataDir: process.env.DATA_DIR || os.tmpdir(),
    // storageType which is your binary package files store. options value is ("local" | "qiniu" | "s3"| "oss" || "tencentcloud")
    storageType: process.env.STORAGE_TYPE || "local",
    // options value is (true | false), when it's true, it will cache updateCheck results in redis.
    updateCheckCache: false,
    // options value is (true | false), when it's true, it will cache rollout results in redis
    rolloutClientUniqueIdCache: false,
  },
  // Config for smtp email，register module need validate user email project source https://github.com/nodemailer/nodemailer
  smtpConfig:{
    host: "smtp.aliyun.com",
    port: 465,
    secure: true,
    auth: {
      user: "",
      pass: ""
    }
  },
  // Config for redis (register module, tryLoginTimes module)
  redis: {
    default: {
      host: "127.0.0.1",
      port: 6379,
      retry_strategy: function (options) {
        if (options.error.code === 'ECONNREFUSED') {
          // End reconnecting on a specific error and flush all commands with a individual error
          return new Error('The server refused the connection');
        }
        if (options.total_retry_time > 1000 * 60 * 60) {
            // End reconnecting after a specific timeout and flush all commands with a individual error
            return new Error('Retry time exhausted');
        }
        if (options.times_connected > 10) {
            // End reconnecting with built in error
            return undefined;
        }
        // reconnect after
        return Math.max(options.attempt * 100, 3000);
      }
    }
  }
}

config.development.log4js = {
  appenders: {console: { type: 'console'}},
  categories : {
    "default": { appenders: ['console'], level:'error'},
    "startup": { appenders: ['console'], level:'info'},
    "http": { appenders: ['console'], level:'info'}
  }
}

config.production = Object.assign({}, config.development);
module.exports = config;


core/utils/qetag.js
// Calculate the eTag of the file. The parameter is buffer or readableStream or file path.
function getEtag(buffer,callback){

	// determine whether the passed parameter is buffer or stream or filepath
	var mode = 'buffer';

	if(typeof buffer === 'string'){
		buffer = require('fs').createReadStream(buffer);
		mode='stream';
	}else if(buffer instanceof require('stream')){
		mode='stream';
	}

	// sha1 algorithm
	var sha1 = function(content){
		var crypto = require('crypto');
		var sha1 = crypto.createHash('sha1');
		sha1.update(content);
		return sha1.digest();
	};

	// Divide by 4M
	var blockSize = 4*1024*1024;
	var sha1String = [];
	var prefix = 0x16;
	var blockCount = 0;

	switch(mode){
		case 'buffer':
			var bufferSize = buffer.length;
			blockCount = Math.ceil(bufferSize / blockSize);

			for(var i=0;i<blockCount;i++){
				sha1String.push(sha1(buffer.slice(i*blockSize,(i+1)*blockSize)));
			}
			process.nextTick(function(){
				callback(calcEtag());
			});
			break;
		case 'stream':
			var stream = buffer;
			stream.on('readable', function() {
				var chunk;
				while (chunk = stream.read(blockSize)) {
					sha1String.push(sha1(chunk));
					blockCount++;
				}
			});
			stream.on('end',function(){
				callback(calcEtag());
			});
			break;
	}

	function calcEtag(){
		if(!sha1String.length){
			return 'Fto5o-5ea0sNMlW_75VgGJCv2AcJ';
		}
		var sha1Buffer = Buffer.concat(sha1String,blockCount * 20);

		// If it is greater than 4M, the sha1 result of each block is sha1 again
		if(blockCount > 1){
			prefix = 0x96;
			sha1Buffer = sha1(sha1Buffer);
		}

		sha1Buffer = Buffer.concat(
			[new Buffer([prefix]),sha1Buffer],
			sha1Buffer.length + 1
		);

		return sha1Buffer.toString('base64')
			.replace(/\//g,'_').replace(/\+/g,'-');

	}

}

module.exports = getEtag;


core/utils/security.js
'use strict';
var bcrypt = require('bcryptjs');
var crypto = require('crypto');
var fs = require('fs');
var Promise = require('bluebird');
var qetag = require('../utils/qetag');
var _ = require('lodash');
var log4js = require('log4js');
var log = log4js.getLogger("cps:utils:security");
var AppError = require('../app-error');

var randToken = require('rand-token').generator({
  chars: '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ',
  source: crypto.randomBytes
});
var security = {};
module.exports = security;

security.md5 = function (str) {
  var md5sum = crypto.createHash('md5');
  md5sum.update(str);
  str = md5sum.digest('hex');
  return str;
}

security.passwordHashSync = function(password){
  return bcrypt.hashSync(password, bcrypt.genSaltSync(12));
}

security.passwordVerifySync = function(password, hash){
  return bcrypt.compareSync(password, hash)
}

security.randToken = function(num) {
  return randToken.generate(num);
}

security.parseToken = function(token) {
  return {identical: token.substr(-9,9), token:token.substr(0,28)}
}

security.fileSha256 = function (file) {
  return new Promise((resolve, reject) => {
    var rs = fs.createReadStream(file);
    var hash = crypto.createHash('sha256');
    rs.on('data', hash.update.bind(hash));
    rs.on('error', (e) => {
      reject(e);
    });
    rs.on('end', () => {
      resolve(hash.digest('hex'));
    });
  });
}

security.stringSha256Sync = function (contents) {
  var sha256 = crypto.createHash('sha256');
  sha256.update(contents);
  return sha256.digest('hex');
}

security.packageHashSync = function (jsonData) {
  var sortedArr = security.sortJsonToArr(jsonData);
  var manifestData = _.filter(sortedArr, (v) => {
    return !security.isPackageHashIgnored(v.path);
  }).map((v) => {
    return v.path + ':' + v.hash;
  });
  log.debug('packageHashSync manifestData:', manifestData);
  var manifestString = JSON.stringify(manifestData.sort());
  manifestString = _.replace(manifestString, /\\\//g, '/');
  log.debug('packageHashSync manifestString:', manifestString);
  return security.stringSha256Sync(manifestString);
}

//The parameter is buffer or readableStream or file path
security.qetag = function (buffer) {
  if (typeof buffer === 'string') {
    try {
      log.debug(`Check upload file ${buffer} fs.R_OK`);
      fs.accessSync(buffer, fs.R_OK);
      log.debug(`Pass upload file ${buffer}`);
    } catch (e) {
      log.error(e);
      return Promise.reject(new AppError.AppError(e.message))
    }
  }
  log.debug(`generate file identical`)
  return new Promise((resolve, reject) => {
    qetag(buffer, (data)=>{
      log.debug('identical:', data);
      resolve(data)
    });
  });
}

security.sha256AllFiles = function (files) {
  return new Promise((resolve, reject) => {
    var results = {};
    var length = files.length;
    var count = 0;
    files.forEach((file) => {
      security.fileSha256(file)
      .then((hash) => {
        results[file] = hash;
        count++;
        if (count == length) {
          resolve(results);
        }
      });
    });
  });
}

security.uploadPackageType = function (directoryPath) {
  return new Promise((resolve, reject) => {
    var recursive = require("recursive-readdir");
    var path = require('path');
    var slash = require("slash");
    recursive(directoryPath, (err, files) => {
      if (err) {
        log.error(new AppError.AppError(err.message));
        reject(new AppError.AppError(err.message));
      } else {
        if (files.length == 0) {
          log.debug(`uploadPackageType empty files`);
          reject(new AppError.AppError("empty files"));
        } else {
          var constName = require('../const');
          const AREGEX=/android\.bundle/
          const AREGEX_IOS=/main\.jsbundle/
          var packageType = 0;
          _.forIn(files, function (value) {
            if (AREGEX.test(value)) {
              packageType = constName.ANDROID;
              return false;
            }
            if (AREGEX_IOS.test(value)) {
              packageType = constName.IOS;
              return false;
            }
          });
          log.debug(`uploadPackageType packageType: ${packageType}`);
          resolve(packageType);
        }
      }
    });
  });
}

// some files are ignored in calc hash in client sdk
// https://github.com/Microsoft/react-native-code-push/pull/974/files#diff-21b650f88429c071b217d46243875987R15
security.isHashIgnored = function (relativePath) {
  if (!relativePath) {
    return true;
  }

  const IgnoreMacOSX = '__MACOSX/';
  const IgnoreDSStore = '.DS_Store';

  return relativePath.startsWith(IgnoreMacOSX)
    || relativePath === IgnoreDSStore
    || relativePath.endsWith(IgnoreDSStore);
}

security.isPackageHashIgnored = function (relativePath) {
  if (!relativePath) {
    return true;
  }

  // .codepushrelease contains code sign JWT
  // it should be ignored in package hash but need to be included in package manifest
  const IgnoreCodePushMetadata = '.codepushrelease';
  return relativePath === IgnoreCodePushMetadata
    || relativePath.endsWith(IgnoreCodePushMetadata)
    || security.isHashIgnored(relativePath);
}


security.calcAllFileSha256 = function (directoryPath) {
  return new Promise((resolve, reject) => {
    var recursive = require("recursive-readdir");
    var path = require('path');
    var slash = require("slash");
    recursive(directoryPath, (error, files) => {
      if (error) {
        log.error(error);
        reject(new AppError.AppError(error.message));
      } else {
        // filter files that should be ignored
        files = files.filter((file) => {
          var relative = path.relative(directoryPath, file);
          return !security.isHashIgnored(relative);
        });

        if (files.length == 0) {
          log.debug(`calcAllFileSha256 empty files in directoryPath:`, directoryPath);
          reject(new AppError.AppError("empty files"));
        }else {
          security.sha256AllFiles(files)
          .then((results) => {
            var data = {};
            _.forIn(results, (value, key) => {
              var relativePath = path.relative(directoryPath, key);
              var matchresult = relativePath.match(/(\/|\\).*/);
              if (matchresult) {
                  relativePath = path.join('CodePush', matchresult[0]);
              }
              relativePath = slash(relativePath);
              data[relativePath] = value;
            });
            log.debug(`calcAllFileSha256 files:`, data);
            resolve(data);
          });
        }
      }
    });
  });
}

security.sortJsonToArr = function (json) {
  var rs = [];
  _.forIn(json, (value, key) => {
    rs.push({path:key, hash: value})
  });
  return _.sortBy(rs, (o) => o.path);
}


config/config.testwin.js
var os = require('os');

var config = {};
config.test = {
  db: {
    username: "root",
    password: "Password12!",
    database: "codepush_test",
    host: "127.0.0.1",
    port: 3306,
    dialect: "mysql",
    logging: false,
    operatorsAliases: false,
  },
  local: {
    storageDir: os.tmpdir(),
    downloadUrl: "http://127.0.0.1:9027/download",
    public: '/download'
  },
  jwt: {
    tokenSecret: 'INSERT_RANDOM_TOKEN_KEY'
  },
  common: {
    tryLoginTimes: 10,
    diffNums: 3,
    dataDir: os.tmpdir(),
    storageType: "local",
    updateCheckCache: true,
    rolloutClientUniqueIdCache: false,
  },
  smtpConfig: false,
  redis: {
    default: {
      host: "127.0.0.1",
      port: 6379,
      retry_strategy: function (options) {
        if (options.error.code === 'ECONNREFUSED') {
          return new Error('The server refused the connection');
        }
        if (options.total_retry_time > 1000 * 60 * 60) {
            return new Error('Retry time exhausted');
        }
        if (options.times_connected > 10) {
            return undefined;
        }
        // reconnect after
        return Math.max(options.attempt * 100, 3000);
      }
    }
  }
}
config.test.log4js = {
  appenders: {console: { type: 'console'}},
  categories : {
    "default": { appenders: ['console'], level:'error'},
    "startup": { appenders: ['console'], level:'info'},
    "http": { appenders: ['console'], level:'info'}
  }
}
module.exports = config;
